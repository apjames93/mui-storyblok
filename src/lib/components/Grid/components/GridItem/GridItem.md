MuiGridItem available components:
* MuiButtonRedirect
* MuiCard
* MuiTypography
* MuiMenu
* MuiIcon
* MuiList
* MuiTable
* MuiPaginationTable
* MuiContactButton
* BlokForm
* MuiImage
* MuiButtonHref
* MuiAudioPlayer
* GoogleConversions
* MuiAccordion