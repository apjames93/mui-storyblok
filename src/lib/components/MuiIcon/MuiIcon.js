import React, { createElement } from 'react';
import PropTypes from 'prop-types';
import Icon from '@material-ui/core/Icon';
import Storyblok from '../../utils/Storyblok';
import { muiStringProp } from '../../utils/customProps';
import BBBSvg from './components/BBBSvg/BBBSvg';
import FacebookSvg from './components/FacebookSvg/FacebookSvg';
import GoogleSvg from './components/GoogleSvg/GoogleSvg';
import LinkedinSvg from './components/LinkedinSvg/LinkedinSvg';
// import styles for icons in styleguide
import '../../styles/styles.scss';

// import icons to use for custom icons passed down to app;
import customIcons from '../../utils/customIcons';

// import { icons } from '../MuiStoryblok/MuiStoryblok';
// const icons = getIcons();
/**
 * MuiIcon is used in storyblok redirect to react routes
 */

export const MuiIcon = ({
  color,
  fontSize,
  iconName,
  rootClass,
}) => {
  const styles = Storyblok.arrayToMuiStyles(rootClass);

  const customIcon = customIcons.find(icon => icon.iconName === iconName);
  if (customIcon) {
    const Custom = customIcon.Component;
    return (
      <Custom
        className={styles.root}
        color={color}
        fontSize={fontSize}
        {...customIcon.props}
      />
    );
  }

  switch (iconName) {
    case 'facebook':
      return <FacebookSvg color={color} fontSize={fontSize} className={styles.root} />;
    case 'bbb':
      return <BBBSvg color={color} fontSize={fontSize} className={styles.root} />;
    case 'linkedin':
      return <LinkedinSvg color={color} fontSize={fontSize} className={styles.root} />;
    case 'google':
      return <GoogleSvg color={color} fontSize={fontSize} className={styles.root} />;
    default:
      return (
        <Icon
          className={styles.root}
          color={color}
          fontSize={fontSize}
        >
          {iconName}
        </Icon>
      );
  }
};

export default MuiIcon;

MuiIcon.propTypes = {
  /** any icon from https://material.io/resources/icons/?style=baseline */
  iconName: PropTypes.string.isRequired,
  /**
   * mui prop: 'inherit', 'primary', 'secondary', 'action', 'error', 'disabled'
   * The color of the component. It supports those theme colors that make sense for this component.
   * */
  color(props, propName, componentName) {
    const validProp = ['inherit', 'primary', 'secondary', 'action', 'error', 'disabled'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * mui prop: 'default', 'small', 'inherit', 'large'
   * The fontSize applied to the icon. Defaults to 24px, but can be configure to inherit font size.
   * */
  fontSize(props, propName, componentName) {
    const validProp = ['default', 'small', 'inherit', 'large'];
    return muiStringProp(props, propName, componentName, validProp);
  },
  /**
   * storyblok multiselect of css classes
   * Override or extend the styles applied to the component
   * */
  rootClass: PropTypes.arrayOf(PropTypes.string),
};

MuiIcon.defaultProps = {
  rootClass: [],
  color: 'secondary',
  fontSize: 'default',
};
