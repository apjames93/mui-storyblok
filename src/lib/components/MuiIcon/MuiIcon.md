MuiIcon example:
[View All MUI Icons](https://material.io/resources/icons/?style=baseline)

```js
  <div styles={{ display: 'flex', flexDirection: 'row' }}>
    <MuiIcon
      iconName='home'
      color='primary'
    />
    <MuiIcon
      iconName='menu'
      color='secondary'
    />
    <MuiIcon
      iconName='close'
      color='action'
    />
    <MuiIcon
      iconName='accessibility'
      color='inherit'
    />
    <MuiIcon
      iconName='done'
      color='error'
    />

      <MuiIcon
      iconName='done'
      color='error'
      fontSize="small"
    />
  </div>
```

Social Icons for MuiIcon examples:
* iconName = 'facebook'
* iconName = 'bbb'
* iconName = 'google'
* iconName = 'linkedin'
* iconName = 'youtube'
* iconName = 'instagram'
* iconName = 'twitter'
* iconName = 'yelp'
* iconName = 'github'

```js
  <div styles={{ display: 'flex', flexDirection: 'row' }}>
    <MuiIcon
      iconName='facebook'
      color='primary'
      fontSize='large'
    />
    <MuiIcon
      iconName='bbb'
      color='secondary'
      fontSize='large'
    />
    <MuiIcon
      iconName='google'
      color='action'
      fontSize='large'
    />
    <MuiIcon
      iconName='linkedin'
      color='inherit'
      fontSize='large'
      rootClass={['margin-left: 10px', 'margin-right: 10px']}
    />
    <MuiIcon
      iconName='youtube'
      color='inherit'
      fontSize='large'
    />
    <MuiIcon
      iconName='instagram'
      color='secondary'
      fontSize='large'
    />
    <MuiIcon
      iconName='twitter'
      color='primary'
      fontSize='large'
    />
    <MuiIcon
      iconName='yelp'
      color='secondary'
      fontSize='large'
    />
    <MuiIcon
      iconName='github'
      color='inherit'
      fontSize='large'
    />
  </div>
```