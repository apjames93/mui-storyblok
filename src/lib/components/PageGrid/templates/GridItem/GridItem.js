import React, { Suspense } from 'react';
import PropTypes from 'prop-types';
import Grid from '@material-ui/core/Grid';
import sizeGrid from 'lib/utils/sizeGrid';
import Storyblok from 'lib/utils/Storyblok';
import { muiStringProp, muiGridProp, muiBlokNumberProp } from 'lib/utils/customProps';
import { renderComponentsWithBridge } from 'lib/utils/customComponents';
import { Box, Hidden } from '@material-ui/core';
import { useInView } from 'react-intersection-observer';

const GridItem = ({
  components,
  alignContent,
  alignItems,
  rootClass,
  direction,
  justify,
  lg,
  md,
  sm,
  wrap,
  spacing,
  xs,
  xl,
  content,
  dataBlokC,
  dataBlokUid,
  storyblokClass,
  transition,
  only,
  backgroundImageUrl,
}) => {
  let heroClass;
  const { ref, inView } = useInView({ threshold: 0.2, triggerOnce: true });
  const gridClass = window?.Storyblok?.inEditor ? {
    borderStyle: 'solid',
    borderColor: '#3889FF',
    borderWidth: '.1em',
  } : {};


  if (backgroundImageUrl) {
    heroClass = {
      ...{
        backgroundImage: `url(${backgroundImageUrl})`,
        backgroundPosition: 'center',
        backgroundRepeat: 'no-repeat',
        backgroundSize: 'cover',
        position: 'relative',
        padding: 0,
        margin: 0,
      },
    };
  }

  const styles = Storyblok.arrayToMuiStyles(rootClass, { ...heroClass });

  return (
    <Hidden only={only}>
      <Grid
        item
        container
        alignContent={alignContent}
        alignItems={alignItems}
        direction={direction}
        justify={justify}
        wrap={wrap}
        spacing={Number(spacing)}
        xs={sizeGrid(xs)}
        sm={sizeGrid(sm)}
        md={sizeGrid(md)}
        lg={sizeGrid(lg)}
        xl={sizeGrid(xl)}
        data-blok-c={dataBlokC}
        data-blok-uid={dataBlokUid}
        className={`${styles.root} ${storyblokClass} ${inView && transition}`}
        style={{ ...gridClass, opacity: inView ? 1 : 0 }}
        inView={inView}
        ref={ref}
      >
        {!content.length && <Box minHeight={200} width={{ xs: '100%' }} />}
        {content.length > 0
          && content.map((component, key) => (
            <Suspense fallback={<></>} key={key}>
              {renderComponentsWithBridge({ ...components }, {
                ...component,
                components,
                key,
              }, key)}
            </Suspense>
          ))
        }
      </Grid>
    </Hidden>
  );
};

export default GridItem;

GridItem.propTypes = {
  /**
   * storyblok multiselect of css classes
   * Mui Override or extend the styles applied to the component.
   */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /**
   * mui prop: 'stretch', 'center', 'flex-start', 'flex-end', 'space-between', 'space-around'
   * Defines the align-content style property. It's applied for all screen sizes.
   */
  alignContent(props, propName, componentName) {
    const validProps = ['stretch', 'center', 'flex-start', 'flex-end', 'space-between', 'space-around'];
    return muiStringProp(props, propName, componentName, validProps);
  },
  /**
   * mui prop: 'flex-start' , 'center' , 'flex-end' , 'stretch' , 'baseline'
   * Defines the align-items style property. It's applied for all screen sizes.
   */
  alignItems(props, propName, componentName) {
    const validProps = ['flex-start', 'center', 'flex-end', 'stretch', 'baseline'];
    return muiStringProp(props, propName, componentName, validProps);
  },
  /*
  * mui prop: 'row' , 'row-reverse' , 'column' , 'column-reverse'
  * Defines the flex-direction style property. It is applied for all screen sizes.
  */
  direction(props, propName, componentName) {
    const validProps = ['row', 'row-reverse', 'column', 'column-reverse'];
    return muiStringProp(props, propName, componentName, validProps);
  },
  /**
   * mui prop:
   'flex-start' , 'center' , 'flex-end' , 'space-between' , 'space-around' , 'space-evenly'
   * Defines the justify-content style property. It is applied for all screen sizes.
   */
  justify(props, propName, componentName) {
    const validProps = ['flex-start', 'center', 'flex-end', 'space-between', 'space-around', 'space-evenly'];
    return muiStringProp(props, propName, componentName, validProps);
  },
  /**
   * mui prop:
   false , 'auto' , true , 1 , 2 , 3 , 4 , 5 , 6 , 7 , 8 , 9 , 10 , 11 , 12
   * Defines the number of grids the component is going to use.
    It's applied for the breakpoint and wider screens if not overridden.
   */
  lg(props, propName, componentName) {
    return muiGridProp(props, propName, componentName);
  },
  /**
   * mui prop: false , 'auto' , true , 1 , 2 , 3 , 4 , 5 , 6 , 7 , 8 , 9 , 10 , 11 , 12
   * Defines the number of grids the component is going to use.
   It's applied for the breakpoint and wider screens if not overridden.
   */
  md(props, propName, componentName) {
    return muiGridProp(props, propName, componentName);
  },
  /**
   * mui prop: false , 'auto' , true , 1 , 2 , 3 , 4 , 5 , 6 , 7 , 8 , 9 , 10 , 11 , 12
   * Defines the number of grids the component is going to use.
   It's applied for the breakpoint and wider screens if not overridden.
   */
  sm(props, propName, componentName) {
    return muiGridProp(props, propName, componentName);
  },
  /**
   * mui prop:  0 , 1 , 2 , 3 , 4 , 5 , 6 , 7 , 8 , 9 , 10
   * Defines the space between the type item component.
   * It can only be used on a type container component.
   */
  // spacing: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  spacing(props, propName, componentName) {
    const validProps = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    return muiBlokNumberProp(props, propName, componentName, validProps);
  },
  /**
   * mui prop: 'nowrap' , 'wrap' , 'wrap-reverse'
   * Defines the flex-wrap style property. It's applied for all screen sizes.
   */
  wrap(props, propName, componentName) {
    const validProps = ['nowrap', 'wrap', 'wrap-reverse'];
    return muiStringProp(props, propName, componentName, validProps);
  },
  /**
   * mui prop: false , 'auto' , true , 1 , 2 , 3 , 4 , 5 , 6 , 7 , 8 , 9 , 10 , 11 , 12
   * Defines the number of grids the component is going to use.
   It's applied for the breakpoint and wider screens if not overridden.
   */
  xs(props, propName, componentName) {
    return muiGridProp(props, propName, componentName);
  },
  /**
   * mui prop: false , 'auto' , true , 1 , 2 , 3 , 4 , 5 , 6 , 7 , 8 , 9 , 10 , 11 , 12
   * Defines the number of grids the component is going to use.
   It's applied for the breakpoint and wider screens if not overridden.
   */
  xl(props, propName, componentName) {
    return muiGridProp(props, propName, componentName);
  },

  /**
   * Content passed to from api
   * can also render any customCompnent passed in
   * will render any component in components prop
   */
  content: PropTypes.arrayOf(PropTypes.shape({
    component: PropTypes.string.isRequired,
  })).isRequired,

  /** storyblok prop for when in editor to allow click bridge */
  dataBlokC: PropTypes.string,
  /** storyblok prop for when in editor to allow click bridge */
  dataBlokUid: PropTypes.string,
  /** storyblok prop for when in editor to allow click bridge */
  storyblokClass: PropTypes.string,
  /** Transition desired to apply on grid item. */
  transition: PropTypes.string,
  /**  components to render in the GridItem */
  components: PropTypes.shape(),
  /**
   * mui prop array of: 'xs' | 'sm' | 'md' | 'lg' | 'xl'
   * Hide the given breakpoint(s).
   * */
  only: PropTypes.arrayOf(PropTypes.string),
  /** url for background img */
  backgroundImageUrl: PropTypes.string,
};

GridItem.defaultProps = {
  alignContent: 'stretch',
  alignItems: 'center',
  rootClass: [],
  direction: 'row',
  justify: 'center',
  lg: false,
  md: false,
  sm: false,
  spacing: '0',
  wrap: 'wrap',
  xl: false,
  xs: false,
  dataBlokC: '',
  dataBlokUid: '',
  storyblokClass: '',
  components: {},
  transition: '',
  only: [],
  backgroundImageUrl: '',
};
