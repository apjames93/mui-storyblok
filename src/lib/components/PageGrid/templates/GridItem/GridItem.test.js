import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import MuiIcon from '../../../MuiIcon/MuiIcon';
import GridItem from './GridItem';

function setup() {
  const props = {
    _editable: '<!--#storyblok#{"name": "GridItem", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
    sizeGrid: jest.fn(),
    components: {
      MuiIcon,
    },
    content: [{
      _editable: '<!--#storyblok#{"name": "Typography", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
      component: 'Typography',
      content: [{
        _editable: '<!--#storyblok#{"name": "Span", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
        component: 'Span',
        content: 'text',
      }],
    }],
  };
  const comp = shallow(<GridItem {...props} />);
  return { comp, props };
}

describe('<GridItem />', () => {
  it('renders GridItem', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();

    const tree = renderer.create((<GridItem {...props} />));
    expect(tree).toMatchSnapshot();
  });
});
