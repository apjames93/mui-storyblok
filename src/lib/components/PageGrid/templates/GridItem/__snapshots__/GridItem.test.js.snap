// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridItem /> snapshot 1`] = `
<div
  className="MuiGrid-root makeStyles-root-3  MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  data-blok-c=""
  data-blok-uid=""
>
  <div
    className="MuiTypography-root makeStyles-root-4 Typography storyblok--outline MuiTypography-body1"
    data-blok-c="{\\"name\\": \\"Typography\\", \\"space\\": \\"48408\\", \\"uid\\": \\"947be9f0-47c3-4315-a95a-550f0c560eb5\\", \\"id\\": \\"307934\\"}"
    data-blok-uid=""
  />
</div>
`;
